pulp_ansible/app/galaxy/v3/views.py

```diff
@@ -108,52 +125,92 @@ class CollectionViewSet(
+
     def get_queryset(self):
         """
-        Returns a CollectionVersions queryset for specified distribution.
+        Returns a Collections queryset for specified distribution.
         """
-        distro_content = self.get_distro_content(self.kwargs["path"])
+        repo_version = self.get_repository_version(self.kwargs["path"])
+        deprecated_query = AnsibleCollectionDeprecated.objects.filter(
+            collection=OuterRef("pk"), repository_version=repo_version
+        )
+        collections = Collection.objects.filter(versions__in=repo_version.content).distinct()
+        collections = collections.annotate(deprecated=Exists(deprecated_query))
 
-        versions = CollectionVersion.objects.filter(pk__in=distro_content).values_list(
+        versions_qs = CollectionVersion.objects.filter(pk__in=repo_version.content).values_list(
             "collection_id",
+            "namespace",
+            "name",
             "version",
         )
 
-        collection_versions = {}
-        for collection_id, version in versions:
-            value = collection_versions.get(str(collection_id))
-            if not value or semantic_version.Version(version) > semantic_version.Version(value):
-                collection_versions[str(collection_id)] = version
-
-        if not collection_versions.items():
-            return CollectionVersion.objects.none()
-
-        query_params = Q()
-        for collection_id, version in collection_versions.items():
-            query_params |= Q(collection_id=collection_id, version=version)
+        highest_versions = defaultdict(
+            lambda: CollectionTuple(None, None, semantic_version.Version("0.0.0"))
+        )
+        for collection_id, namespace, name, version in versions_qs:
+            possible_highest = semantic_version.Version(version)
+            if possible_highest > highest_versions[collection_id].version:
+                highest_versions[collection_id] = CollectionTuple(namespace, name, possible_highest)
 
-        collections = CollectionVersion.objects.select_related("collection").filter(query_params)
+        self.highest_versions_context = highest_versions  # needed by get__serializer_context
         return collections
 
     def get_object(self):
         """
         Returns a Collection object.
         """
         queryset = self.filter_queryset(self.get_queryset())
-
         return get_object_or_404(
             queryset, namespace=self.kwargs["namespace"], name=self.kwargs["name"]
         )
 
-    def update(self, request, *args, **kwargs):
+    def get_serializer_context(self, *args, **kwargs):
+        """
+        Return the serializer context.
+
+        This uses super() but also adds in the "highest_versions" data from get_queryset()
+        """
+        super_data = super().get_serializer_context()
+        super_data["highest_versions"] = self.highest_versions_context
+        return super_data
+
+    def update(self, request):
```

Comment:
A quick check of `curl -X PATCH http://localhost:5001/pulp_ansible/galaxy/published/api/v3/collections/awcrosby/collection_test/ deprecated=true` yields `TypeError at /pulp_ansible/galaxy/published/api/v3/collections/awcrosby/collection_test/ update() got an unexpected keyword argument 'path'`... I'll look more closely at it tomorrow

Reply:
The 'path' argument comes from "repository" name in the URL, let's talk on slack to work through it maybe.

Reply:
It comes from here https://github.com/pulp/pulp_ansible/blob/master/pulp_ansible/app/urls.py#L18

Reply:
If you share the specific traceback you get I can maybe help sort it out.

Reply:
@bmbouter here is the traceback: https://gist.github.com/awcrosby/dfcb47c0e11a98310494514b818dc948

Reply:
updating pulp_ansible `update(self, request)` with `update(self, request, *args, **kwargs)` get beyond the initial error

Reply:
What API call triggers this?

Reply:
api call: `curl -X PATCH http://localhost:5001/pulp_ansible/galaxy/published/api/v3/collections/awcrosby/collection_test/ deprecated=true`

with the edit ^ to update(), currently seeing this:
```
File "/src/pulp_ansible/pulp_ansible/app/galaxy/v3/views.py", line 206, in update
deprecated_value = serializer.validated_data["deprecated"]
KeyError: 'deprecated'
```

Reply:
@awcrosby how do you create the data up to that point? Here's the script I run:  https://paste.centos.org/view/b6bdcb9a

And then I make a call like `http PATCH http://pulp3-source-fedora32.localhost.example.com/pulp_ansible/galaxy/my_content/api/v3/collections/testing/k8s_demo_collection/ deprecated=false`

Reply:
Oh interesting, when I try `curl -X PATCH ...` I also get the `TypeError: update() got an unexpected keyword argument 'path'`

Reply:
I set up the data via an import to galaxy_ng

Reply:
httpie gives a `unexpected keyword argument 'path'` error also: 
https://gist.github.com/awcrosby/4d1908d4ae1782321eba6c6da4265d10

Reply:
with the latest change this patch is working!

Reply:
Awww yisssss